File: src/parser.ts
/**
 * parser.ts — Markdown task parser for Obsidian Tasks plugin syntax.
 *
 * Parses task lines from vault markdown files.
 * Supports inline fields: [type:: ...], [refs:: ...]
 * Attaches indented `comment::` subtopics to their parent task.
 */

import { sanitizeForDisplay } from "./sanitize.ts";

// ── Types ─────────────────────────────────────────────────────────────────────

export type TaskStatus = "todo" | "in_progress" | "done" | "cancelled" | "backlog";

export interface TaskComment {
  text: string;
  lineNumber: number;
}

export interface Task {
  /** Display text (inline fields stripped, sanitized) */
  text: string;
  status: TaskStatus;
  type?: string;
  refs?: string;
  comments: TaskComment[];
  sourceFile: string;
  lineNumber: number;
  /**
   * Stable identity key for safe mutations.
   * Format: "<sourceFile>:L<lineNumber>:<normalizedText>"
   */
  fingerprint: string;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function toStatus(checkChar: string): TaskStatus {
  switch (checkChar) {
    case " ": return "todo";
    case "/": return "in_progress";
    case "x":
    case "X": return "done";
    case "-": return "cancelled";
    case "b": return "backlog";
    default:  return "todo";
  }
}

const RE_FIELD_TYPE = /\[type::\s*([^\]]+)\]/;
const RE_FIELD_REFS = /\[refs::\s*([^\]]+)\]/;

function extractField(text: string, field: "type" | "refs"): string | undefined {
  const re = field === "type" ? RE_FIELD_TYPE : RE_FIELD_REFS;
  const m = text.match(re);
  return m ? sanitizeForDisplay(m[1]!.trim(), 100) : undefined;
}

function stripFields(text: string): string {
  return text.replace(/\[[^\]]+::\s*[^\]]+\]/g, "").replace(/\s+/g, " ").trim();
}

function buildFingerprint(sourceFile: string, lineNumber: number, normalizedText: string): string {
  return `${sourceFile}:L${lineNumber}:${normalizedText}`;
}

function parseCommentLine(line: string, lineNumber: number): TaskComment | null {
  const m = line.match(/^(\s{2,})[-*]\s+comment::\s+(.+)$/);
  if (!m) return null;
  const text = sanitizeForDisplay(m[2]!.trim(), 200);
  if (!text) return null;
  return { text, lineNumber };
}

// ── Public API ────────────────────────────────────────────────────────────────

export function parseTaskLine(
  line: string,
  sourceFile: string,
  lineNumber: number,
): Task | null {
  const m = line.match(/^\s*[-*]\s+\[([^\]]*)\]\s+(.+)$/);
  if (!m) return null;
  const checkChar = m[1]!;
  if (checkChar.length > 1) return null;

  const rawSanitized = sanitizeForDisplay(m[2]!, 300);
  const displayText = sanitizeForDisplay(stripFields(rawSanitized), 200);

  return {
    text: displayText,
    status: toStatus(checkChar),
    type: extractField(rawSanitized, "type"),
    refs: extractField(rawSanitized, "refs"),
    comments: [],
    sourceFile,
    lineNumber,
    fingerprint: buildFingerprint(sourceFile, lineNumber, displayText),
  };
}

export function parseProjectFile(content: string, sourceFile: string): Task[] {
  const tasks: Task[] = [];
  const lines = content.split("\n");
  let lastTask: Task | null = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]!;
    const lineNumber = i + 1;

    const comment = lastTask ? parseCommentLine(line, lineNumber) : null;
    if (comment) { lastTask!.comments.push(comment); continue; }

    const task = parseTaskLine(line, sourceFile, lineNumber);
    if (task) { tasks.push(task); lastTask = task; continue; }

    if (line.trim() !== "") lastTask = null;
  }

  return tasks;
}

File: internal/parser/parser.go
// Package parser parses Obsidian Tasks plugin task lines from Markdown files.
//
// Supported syntax:
//
//	- [ ]  Todo
//	- [/]  In Progress
//	- [x]  Done  (also [X])
//	- [-]  Cancelled
//
// Inline fields:
//
//	[type:: value]
//	[refs:: value]
//
// Comment subtopics (indented, attached to parent task):
//
//	  - comment:: 2026-04-12 14:30 @user — note text
package parser

import (
	"regexp"
	"strings"

	"github.com/pot-labs/otb/internal/sanitize"
)

// TaskStatus represents the four possible Obsidian Tasks plugin states.
type TaskStatus string

const (
	StatusTodo       TaskStatus = "todo"
	StatusInProgress TaskStatus = "in_progress"
	StatusDone       TaskStatus = "done"
	StatusCancelled  TaskStatus = "cancelled"
	StatusBacklog    TaskStatus = "backlog"
)

// TaskComment is a single comment line attached to a task.
type TaskComment struct {
	Text       string
	LineNumber int
}

// Task is a parsed task with all metadata.
type Task struct {
	Text        string       // display text (inline fields stripped, sanitized)
	Status      TaskStatus
	Type        string       // optional [type:: ...] value
	Refs        string       // optional [refs:: ...] value
	Comments    []TaskComment
	SourceFile  string
	LineNumber  int
	Fingerprint string       // "<sourceFile>:L<line>:<normalizedText>"
}

// Field limits (in runes) — any value beyond these is truncated.
const (
	maxText    = 300
	maxType    = 100
	maxRefs    = 200
	maxComment = 200
)

var (
	reTask    = regexp.MustCompile(`^(\s*)[-*]\s+\[([^\]]*)\]\s+(.+)$`)
	reField   = regexp.MustCompile(`\[[^\]]+::\s*[^\]]+\]`)
	reType    = regexp.MustCompile(`\[type::\s*([^\]]+)\]`)
	reRefs    = regexp.MustCompile(`\[refs::\s*([^\]]+)\]`)
	reComment = regexp.MustCompile(`^(\s{2,})[-*]\s+comment::\s+(.+)$`)
)

func toStatus(ch string) TaskStatus {
	switch ch {
	case "/":
		return StatusInProgress
	case "x", "X":
		return StatusDone
	case "-":
		return StatusCancelled
	case "b":
		return StatusBacklog
	default:
		return StatusTodo
	}
}

func extractField(raw, field string) string {
	var re *regexp.Regexp
	switch field {
	case "type":
		re = reType
	case "refs":
		re = reRefs
	default:
		return ""
	}
	m := re.FindStringSubmatch(raw)
	if m == nil {
		return ""
	}
	limit := maxType
	if field == "refs" {
		limit = maxRefs
	}
	return sanitize.ForDisplay(strings.TrimSpace(m[1]), limit)
}

func stripFields(s string) string {
	return strings.Join(strings.Fields(reField.ReplaceAllString(s, "")), " ")
}

func fingerprint(sourceFile string, lineNumber int, normalizedText string) string {
	return sourceFile + ":L" + itoa(lineNumber) + ":" + normalizedText
}

// itoa is a minimal int-to-string helper to avoid importing fmt.
func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	buf := make([]byte, 0, 10)
	for n > 0 {
		buf = append([]byte{byte('0' + n%10)}, buf...)
		n /= 10
	}
	if neg {
		buf = append([]byte{'-'}, buf...)
	}
	return string(buf)
}

// ParseTaskLine parses a single Markdown line into a Task.
// Returns nil if the line is not a task line.
func ParseTaskLine(line, sourceFile string, lineNumber int) *Task {
	m := reTask.FindStringSubmatch(line)
	if m == nil {
		return nil
	}
	checkChar := m[2]
	// Security: reject multi-rune checkbox values
	if len([]rune(checkChar)) > 1 {
		return nil
	}
	rawContent := m[3]
	sanitizedRaw := sanitize.ForDisplay(rawContent, maxText)
	displayText := sanitize.ForDisplay(stripFields(sanitizedRaw), maxText)

	return &Task{
		Text:        displayText,
		Status:      toStatus(checkChar),
		Type:        extractField(sanitizedRaw, "type"),
		Refs:        extractField(sanitizedRaw, "refs"),
		Comments:    nil,
		SourceFile:  sourceFile,
		LineNumber:  lineNumber,
		Fingerprint: fingerprint(sourceFile, lineNumber, displayText),
	}
}

func parseCommentLine(line string, lineNumber int) *TaskComment {
	m := reComment.FindStringSubmatch(line)
	if m == nil {
		return nil
	}
	text := sanitize.ForDisplay(strings.TrimSpace(m[2]), maxComment)
	if text == "" {
		return nil
	}
	return &TaskComment{Text: text, LineNumber: lineNumber}
}

// ParseProjectFile parses all task lines (with attached comments) from
// the content of a single Markdown file.
func ParseProjectFile(content, sourceFile string) []Task {
	var tasks []Task
	lines := strings.Split(content, "\n")
	var lastTask *Task

	for i, line := range lines {
		lineNumber := i + 1

		if lastTask != nil {
			if cmt := parseCommentLine(line, lineNumber); cmt != nil {
				lastTask.Comments = append(lastTask.Comments, *cmt)
				continue
			}
		}

		if task := ParseTaskLine(line, sourceFile, lineNumber); task != nil {
			tasks = append(tasks, *task)
			lastTask = &tasks[len(tasks)-1]
			continue
		}

		if strings.TrimSpace(line) != "" {
			lastTask = nil
		}
	}

	return tasks
}
